src/migmose/mig/reducednestednachrichtenstruktur.py

```diff
@@ -2,16 +2,135 @@
+from migmose.mig.nestednachrichtenstruktur import NestedNachrichtenstruktur
 
 
 class ReducedNestedNachrichtenstruktur(BaseModel):
     """will contain the tree structure of nachrichtenstruktur tables"""
 
     header_linie: Optional[NachrichtenstrukturZeile] = None
-    segmente: list[NachrichtenstrukturZeile]
-    segmentgruppe: list["ReducedNestedNachrichtenstruktur"]
+    segmente: list[Optional[NachrichtenstrukturZeile]] = []
+    segmentgruppen: list[Optional["ReducedNestedNachrichtenstruktur"]] = []
+
+    @classmethod
+    def create_reduced_nested_nachrichtenstruktur(
+        cls, nachrichten_struktur: NestedNachrichtenstruktur
+    ) -> "ReducedNestedNachrichtenstruktur":
+        """init nested Nachrichtenstruktur"""
+
+        # Helper function to create a unique identifier for each segment
+        def get_identifier(segment: NachrichtenstrukturZeile | None) -> tuple[str, str]:
+            if segment is None:
+                return ("0", "root")
+            return (segment.zaehler, segment.bezeichnung)
+
+        def count_segments(segment_group: ReducedNestedNachrichtenstruktur | NestedNachrichtenstruktur) -> int:
+            # Start with counting segments directly under the current segment group
+            total_segments = len(segment_group.segmente)
+            # Recursively count segments in nested segment groups
+            for nested_sg in segment_group.segmentgruppen:
+                if nested_sg is not None:
+                    total_segments += count_segments(nested_sg)
+
+            return total_segments
+
+        # Function to process segments and remove duplicates within the same list.
+        def process_segments(segments: list[Optional[NachrichtenstrukturZeile]]):
+            seen = set()
+            unique_segments = []
+            for segment in segments:
+                if segment is not None:
+                    identifier = get_identifier(segment)
+                    if identifier not in seen:
+                        seen.add(identifier)
+                        unique_segments.append(segment)
+            return unique_segments
+
+        # Recursive function to traverse and clean segment groups
+        def process_segmentgruppen(
+            segmentgruppen: list[NestedNachrichtenstruktur | None], segment_count_dict: dict, seen=None, depth: int = 0
+        ):
+            """Recursively clean segment groups to avoid duplicates, keep largest,
+            with debugging for circular references."""
+            if seen is None:
+                seen = {}
+            result = []
+
+            for sg in segmentgruppen:
+                if sg is not None:
+                    identifier = get_identifier(sg.header_linie)
+                    max_count, max_sg = segment_count_dict[identifier]
+
+                    if identifier not in seen:
+                        seen[identifier] = max_sg
+                        print(f"Added {identifier} with {max_count} segments at depth {depth}.")
```

Comment:
please avoid printing and use logging instead.

Reply:
https://github.com/Hochfrequenz/migmose/pull/41/commits/9224b81b77bef6471743ea8b34acc13ae59780db